src/arch/riscv/pagetable_walker.cc

```diff
@@ -399,21 +715,55 @@ Walker::WalkerState::stepWalk(PacketPtr &write)
                     }
                     // perform step 8 only if pmp checks pass
                     if (fault == NoFault) {
-                        DPRINTF(PageTableWalker,
-                                "#0 leaf node at level %d, with vpn %#x\n",
-                                 level, entry.vaddr);
-
-                        // step 8
-                        entry.logBytes = PageShift + (level * LEVEL_BITS);
-                        entry.paddr = pte.ppn;
-                        entry.vaddr &= ~((1 << entry.logBytes) - 1);
-                        entry.pte = pte;
-                        // put it non-writable into the TLB to detect
-                        // writes and redo the page table walk in order
-                        // to update the dirty flag.
-                        if (!pte.d && mode != BaseMMU::Write)
-                            entry.pte.w = 0;
-                        doTLBInsert = true;
+                        // TLB inserts are OK for single stage walks
+                        // For two-stage, FIRST_STAGE will reach here just once
+                        // (GStage reaches here multiple times)
+                        if (walkType == OneStage || walkType == GstageOnly ||
+                            (walkType == TwoStage && curstage == FIRST_STAGE))
+                        {
```

Comment:
Style comment: This is *way* too deep of nesting. I suggest adding a function.

Reply:
Completely agree, but I did not want to extract out the logic, especially since all the previous steps are numbered inline with comments.

Still, we can refactor this if it is a problem.